### ISSUE TITLE
Can MMT be trained with EN->ET, EN-LT and other less popular European language pairs?

### ISSUE THREAD

**Issue by mindaugasDEV** [ModernMT/MMT#146]
Hello,

I have compiled MMT as per Install file.
EN-IT engine was created successfully from /examples/data/train files.

However, if I change the file extensions of example target training files from *.it (Italian) to *.lt (for Lithuanian), then 
command 
`./mmt create en lt examples/data/train` freezes in INFO: (2 of 6) Corpora preprocessing and I need to cancel the script.

The same, if I change the endings of files with target training data to et (Estonian), still the script stops in step 2, Corpora preprocessing.

In comparison, if I change the file extensions to lv (for Latvian) (or ru - for Russian, or fi - for Finnish), then script goes via all 6 steps.

Does it mean that at the moment MMT cannot handle Estonian and Lithuanian, only Russian and Latvian from the above language group? If yes, when you plan to add those languages? Does it depend on some script that does not have rules for those languages? Can we help you with that (as those languages are important to us)?

Thanks,
Mindaugas

**Comment by marcotrombetti** [comment]
Good question.

It should work with lt and lv but we never tested those.
We will let you know soon.



2016-06-28 0:43 GMT+02:00 mindaugasDEV <notifications@github.com>:

**Comment by mindaugasDEV** [comment]
Yes, I would be great if you check. Thanks!

LV, RU, FI is working (maybe because Moses has more scripts for grammar of those languages)
ET, LT is NOT WORKING (assumption - official Moses is not well developed for those languages, although commercial MT providers serve those languages conveniently).

**Comment by davidecaroselli** [comment]
The issue is due the fact that ET and LT are not currently supported and this situation is not properly reported by the script (that hangs instead returning an error).

However, we're working on implementing a backoff solution for the languages for which we don't have a custom Tokenizer. I'll let you know as soon as this feature will be implemented.

**Comment by mindaugasDEV** [comment]
Thanks,
Uff, when are you going to implement those languages? Is it a matter of 1-2 months, or 3rd quarter?

Can we then ourselves go for “shortcut solution“:
1) change language code in TMX from ET (unsupported) to FI (supported language with similar grammar structure) or from LT (unsupported) to LV (that has similar grammar structure)
2) then train the FI or LV engine (while in reality they are ET or LT)
3) use engines by calling FI or LV for ET or LT texst.

In other words, use related languages to overcome obstacle of not supported languages. It is not a long term solution, but should work, unless you see problems as tokenizers will behave badly with “similar, but not exact“ language grammar.

Mindaugas

**Comment by davidecaroselli** [comment]
What you are proposing is a perfectly valid solution. Moreover, this behavior is exactly the kind of backoff we want to implement: some tokenizers are not so hyper-specific, and can be safely used also with other similar languages.

Let me know if you find the output of such engine good enough, and we can setup the backoff for those languages very quickly :)

**Comment by mindaugasDEV** [comment]
OK, I will try it after you fix the bug #147 :)

**Comment by davidecaroselli** [comment]
@mindaugasDEV 

Starting from commit https://github.com/ModernMT/MMT/commit/b0d04c6d02c5b3d4b4070de00ae3faee6f72a6b0 MMT has a default Tokenizer implementation that is used when no language-specific implementation is available.

So you can now train MMT on any language pair :)

**Comment by davidecaroselli** [comment]
In theory yes, but with the actual implementation of MMT tokenizers there should be no much differences with a generic implementation. The difference between the generic one and, i.e. FI, is a list of unbreakable words. This list, as you can imagine, is very language specific so it's hardly reusable.

All the generic best-practices for tokenizations are already included in the "generic implementation", for this reason I'm convinced that the best option for now is to use the Standard tokenizer when a more specific one is not available.

**Comment by mindaugasDEV** [comment]
Ok!:)
